bigchaindb/backend/mongodb/connection.py

```diff
@@ -43,32 +60,75 @@ def conn(self):
     def db(self):
         return self.conn[self.dbname]
 
+    def query(self):
+        return Lazy()
+
+    def collection(self, name):
+        """Return a lazy object that can be used to compose a query.
+
+        Args:
+            name (str): the name of the collection to query.
+        """
+        return self.query()[self.dbname][name]
+
+    def run(self, query):
+        try:
+            return query.run(self.conn)
+        except pymongo.errors.DuplicateKeyError as exc:
+            raise DuplicateKeyError from exc
+        except pymongo.errors.AutoReconnect as exc:
+            raise ConnectionError from exc
+        except pymongo.errors.OperationFailure as exc:
+            raise OperationError from exc
+
     def _connect(self):
-        # we should only return a connection if the replica set is
-        # initialized. initialize_replica_set will check if the
-        # replica set is initialized else it will initialize it.
-        initialize_replica_set()
+        """Try to connect to the database.
 
-        for i in range(self.max_tries):
-            try:
-                self.connection = MongoClient(self.host, self.port,
-                                              replicaset=self.replicaset)
-            except errors.ConnectionFailure:
-                if i + 1 == self.max_tries:
-                    raise
-                else:
-                    time.sleep(2**i)
+        Raises:
+            :exc:`~ConnectionError`: If the connection to the database
+                fails.
+        """
 
+        attempt = 0
+        for i in self.max_tries_counter:
+            attempt += 1
 
-def initialize_replica_set():
+            try:
+                # we should only return a connection if the replica set is
+                # initialized. initialize_replica_set will check if the
+                # replica set is initialized else it will initialize it.
+                initialize_replica_set(self.host, self.port, self.connection_timeout)
+
+                # FYI: this might raise a `ServerSelectionTimeoutError`,
+                # that is a subclass of `ConnectionFailure`.
+                self.connection = pymongo.MongoClient(self.host,
+                                                      self.port,
+                                                      replicaset=self.replicaset,
+                                                      serverselectiontimeoutms=self.connection_timeout)
+
+            # `initialize_replica_set` might raise `ConnectionFailure` or `OperationFailure`.
+            except (pymongo.errors.ConnectionFailure,
+                    pymongo.errors.OperationFailure) as exc:
+                logger.warning('Attempt %s/%s. Connection to %s:%s failed after %sms.',
+                               attempt, self.max_tries if self.max_tries != 0 else '∞',
+                               self.host, self.port, self.connection_timeout)
+                if attempt == self.max_tries:
```

Comment:
Can't this be moved outside the for loop?
`if attempt == self.max_tries` seems to be doing the same as `for i in self.max_tries_counter`